subprojects/plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/GenerateScriptPluginAdapters.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.plugins.precompiled
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.FileTree
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+
+import java.io.File
+
+
+@CacheableTask
+open class GenerateScriptPluginAdapters : DefaultTask() {
+
+    @InputFiles
+    @PathSensitive(PathSensitivity.RELATIVE)
+    internal
+    lateinit var scripts: FileTree
+
+    @Internal
+    internal
+    lateinit var plugins: List<ScriptPlugin>
+
+    @OutputDirectory
+    var outputDirectory = project.objects.directoryProperty()
+
+    @TaskAction
+    @Suppress("unused")
+    internal
+    fun generate() =
+        outputDirectory.asFile.get().let { outputDir ->
+            outputDir.mkdirs()
+            for (scriptPlugin in plugins) {
+                scriptPlugin.writeScriptPluginAdapterTo(outputDir)
+            }
+        }
+}
+
+
+internal
+fun ScriptPlugin.writeScriptPluginAdapterTo(outputDir: File) {
+
+    val (packageDir, packageDeclaration) =
+        packageName?.let { packageName ->
+            packageDir(outputDir, packageName) to "package $packageName"
+        } ?: outputDir to ""
+
+    val outputFile =
+        packageDir.resolve("$simplePluginAdapterClassName.kt")
+
+    outputFile.writeText("""
+
+        $packageDeclaration
+
+        /**
+         * Precompiled [$scriptFileName][$compiledScriptTypeName] script plugin.
+         *
+         * @see $compiledScriptTypeName
+         */
+        class $simplePluginAdapterClassName : org.gradle.api.Plugin<$targetType> {
```

Comment:
I can't come up with a concrete example of where this might blow up, but I think to be safe, the name of the class needs to be wrapped in backticks.
```
class `$simplePluginAdapterClassName` : org.gradle.api.Plugin<$targetType> {
```

Reply:
That's not necessary for `simplePluginAdapterClassName` is guaranteed to be a valid Java identifier per https://github.com/gradle/kotlin-dsl/blob/071d76cf5c78736f00c22fd6240ed425df3acf51/subprojects/plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/ScriptPlugin.kt#L59-L63

That assumption could be made clearer with an inline class though.

Reply:
Okay, I think I have an example.
If your script is named, `🦆.gradle.kts`.
A class called `🦆Plugin` will not compile without the backticks.
I'm guessing the same problem exists if you have spaces in the name of your file like `a great name.gradle.kts`.

Reply:
@JLLeitschuh, this should respectively be `_Plugin` and `Great_namePlugin`, have another look at Rodrigo's comment above and see what `asJavaIdentifier()` does.

Reply:
Right, okay. I understand.
A quick test that demonstrates the exact problem you're referring to:
```kotlin
@Test
fun `script name collision test`() {
    val script1 =
        newFile("file#22.gradle.kts")
    val script2 =
        newFile("file,22.gradle.kts")

    val scriptPlugin1 = ScriptPlugin(script1)
    val scriptPlugin2 = ScriptPlugin(script2)

    assertThat(
        scriptPlugin1.simplePluginAdapterClassName,
        not(equalTo(scriptPlugin2.simplePluginAdapterClassName)))
}
```
This currently fails because the two class names are the same.

Reply:
@eskatos Followup here: https://github.com/gradle/kotlin-dsl/issues/1284